src/box/xlog.c

```diff
@@ -1158,43 +1166,68 @@ xlog_tx_write_zstd(struct xlog *log)
 	}
 	uint32_t crc32c = 0;
 	struct iovec *iov;
-	/* 3 is compression level. */
-	ZSTD_compressBegin(log->zctx, 3);
 	size_t offset = XLOG_FIXHEADER_SIZE;
 	for (iov = log->obuf.iov; iov->iov_len; ++iov) {
-		/* Estimate max output buffer size. */
-		size_t zmax_size = ZSTD_compressBound(iov->iov_len - offset);
-		/* Allocate a destination buffer. */
-		void *zdst = obuf_reserve(&log->zbuf, zmax_size);
-		if (!zdst) {
-			diag_set(OutOfMemory, zmax_size, "runtime arena",
-				  "compression buffer");
-			goto error;
-		}
-		size_t (*fcompress)(ZSTD_CCtx *, void *, size_t,
-				    const void *, size_t);
+		ZSTD_outBuffer zobuf = {
+			.dst = NULL,
+			.size = 0,
+			.pos = 0,
+		};
+		ZSTD_inBuffer zibuf = {
+			.src = (char *)iov->iov_base + offset,
+			.size = iov->iov_len - offset,
+			.pos = 0,
+		};
 		/*
 		 * If it's the last iov or the last
 		 * log has 0 bytes, end the stream.
 		 */
-		if (iov == log->obuf.iov + log->obuf.pos ||
-		    !(iov + 1)->iov_len) {
-			fcompress = ZSTD_compressEnd;
-		} else {
-			fcompress = ZSTD_compressContinue;
-		}
-		size_t zsize = fcompress(log->zctx, zdst, zmax_size,
-					 (char *)iov->iov_base + offset,
-					 iov->iov_len - offset);
-		if (ZSTD_isError(zsize)) {
-			diag_set(ClientError, ER_COMPRESSION,
-				 ZSTD_getErrorName(zsize));
-			goto error;
+		ZSTD_EndDirective mode =
+			iov == log->obuf.iov + log->obuf.pos ||
+			(iov + 1)->iov_len == 0 ? ZSTD_e_end : ZSTD_e_continue;
+		while (true) {
+			if (zobuf.dst != NULL) {
+				/*
+				 * Advance output buffer to the end of
+				 * compressed data.
+				 */
+				obuf_alloc(&log->zbuf, zobuf.pos);
+				/* Update crc32c */
+				crc32c = crc32_calc(crc32c, zobuf.dst,
+						    zobuf.pos);
+			}
+			size_t remaining = zibuf.size - zibuf.pos;
+			/*
+			 * `ZSTD_compressBound` fails, if  provided
+			 * size > `ZSTD_MAX_INPUT_SIZE`.
+			 */
+			zobuf.size = MIN(ZSTD_MAX_INPUT_SIZE,
+					 MAX(64, /* `remaining == 0` */
```

Comment:
Nit: it would be nice to elaborate on this constant more (I would move it to a separate constant like `min_buf_size` and write that we are concerned about progress guarantee, so picked a constant which is equal to `compressBound(0)` in the current version of `zstd`).

Reply:
Done.